### PR TITLE
MDMA Offload for camera data capture.

### PR DESCRIPTION
**Pull request by kwagyeman**
Adds support for MDMA data movement of lines from the DMA line buffer to the image buffer for CPU offload.

### CODE REVIEW

**Comment by iabdalkader** [comment]
Have a few improvements:
1) We can merge the two paths in  `void DCMI_DMAConvCpltUser(uint32_t addr)` and replace with this:
If you'd like to avoid branches in unaligned_memcpy, we can add new set of functions `unaligned_memcpy_colwise_xx`

```
switch (sensor.pixformat) {
    case PIXFORMAT_BAYER:
        dst += buffer->offset * (sensor.transpose) ? 1 : MAIN_FB()->u;
        src += MAIN_FB()->x;
        #if (OMV_ENABLE_SENSOR_MDMA == 1)
        mdma_memcpy(buffer, dst, src, sizeof(uint8_t), sensor.transpose);
        #else
        unaligned_memcpy(dst, src, MAIN_FB()->u, MAIN_FB()->v, sensor.transpose);
        #endif
        break;
    case PIXFORMAT_GRAYSCALE:
        dst += buffer->offset * (sensor.transpose) ? 1 : MAIN_FB()->u;
        if (sensor.gs_bpp == sizeof(uint8_t)) {
            // 1BPP GRAYSCALE.
            src += MAIN_FB()->x;
            #if (OMV_ENABLE_SENSOR_MDMA == 1)
            mdma_memcpy(buffer, dst, src, sizeof(uint8_t), sensor.transpose);
            #else
            unaligned_memcpy(dst, src, MAIN_FB()->u, MAIN_FB()->v, sensor.transpose);
            #endif
        } else {
            // Extract Y channel from YUV.
            src16 += MAIN_FB()->x;
            #if (OMV_ENABLE_SENSOR_MDMA == 1)
            mdma_memcpy(buffer, dst, src16, sizeof(uint8_t), sensor.transpose);
            #else
            unaligned_2_to_1_memcpy(dst, src16, MAIN_FB()->u, MAIN_FB()->v, sensor.transpose);
            #endif
        }
        break;
    case PIXFORMAT_RGB565:
    case PIXFORMAT_YUV422:
        dst16 += buffer->offset * (sensor.transpose) ? 1 : MAIN_FB()->u;
        src16 += MAIN_FB()->x;
        #if (OMV_ENABLE_SENSOR_MDMA == 1)
        mdma_memcpy(buffer, dst16, src16, sizeof(uint16_t), sensor.transpose);
        #else
        if (SENSOR_HW_FLAGS_GET(&sensor, SWNSOR_HW_FLAGS_RGB565_REV)) {
            unaligned_memcpy_rev16(dst16, src16, MAIN_FB()->u, MAIN_FB()->v, sensor.transpose);
        } else {
            unaligned_memcpy(dst16, src16, MAIN_FB()->u * sizeof(uint16_t), MAIN_FB()->v, sensor.transpose);
        }
        #endif
        break;
    default:
        break;
}
```

2) In snapshot that's really a lot to do in one function, all the MDMA setup should be moved to a new function.

```
#if (OMV_ENABLE_SENSOR_MDMA == 1)
int mdma_config(sensor_t *sensor, uint32_t line_width_bytes, bytes_per_pixel, etc...)
{
}
#endif
```
3) The long if else chains (for source/dest inc, size burst etc..) should be replaced by either this (note the size is just bits from 0->n shifted by Pos):
```
DCMI_MDMA_Handle0.Init.SourceBurst  = MDMA_SOURCE_BURST_SINGLE;
DCMI_MDMA_Handle0.Init.DestBurst    = MDMA_DEST_BURST_SINGLE;
for (uint32_t i=7; i>0; i--) {
    if ((buffer_transfer_burst_size % (2<<(i-1))) == 0) {
        DCMI_MDMA_Handle0.Init.SourceBurst  = i << MDMA_CTCR_SBURST_Pos;
        DCMI_MDMA_Handle0.Init.DestBurst    = i << MDMA_CTCR_DBURST_Pos;
        break;
    }
}
```
Or just const lookup tables:
```
static const uint32_t MDMA_BURST_SIZE[][2] = {
         {MDMA_SOURCE_BURST_SINGLE, MDMA_DEST_BURST_SINGLE},
         {MDMA_SOURCE_BURST_2BEATS, MDMA_DEST_BURST_2BEATS},
         .....
};

DCMI_MDMA_Handle0.Init.SourceBurst  = MDMA_BURST_SIZE[0][0];
DCMI_MDMA_Handle0.Init.DestBurst      = MDMA_BURST_SIZE[0][1];
for (uint32_t i=7; i>0; i--) {
    if ((buffer_transfer_burst_size % (2<<(i-1))) == 0) {
        DCMI_MDMA_Handle0.Init.SourceBurst  = MDMA_BURST_SIZE[i][0];
        DCMI_MDMA_Handle0.Init.DestBurst      = MDMA_BURST_SIZE[i][1];
        break;
    }
}
```

**Comment by kwagyeman** [comment]
Okay, I've done 2 and 3. Just need to redo 1.

**Comment by kwagyeman** [comment]
Done with everything. Just need to test on all cameras with all camera sensors.

**Comment by kwagyeman** [comment]
Checked all cameras. This is ready to be merged.

File: src/omv/ports/stm32/sensor.c [hunk 1]
```diff
@@ -30,12 +30,20 @@
...
+#define DMA_LENGTH_ALIGNMENT    (16)
+#define SENSOR_TIMEOUT_MS       (3000)
+
+sensor_t sensor = {};
+static TIM_HandleTypeDef  TIMHandle  = {.Instance = DCMI_TIM};
+static DMA_HandleTypeDef  DMAHandle  = {.Instance = DMA2_Stream1};
+static DCMI_HandleTypeDef DCMIHandle = {.Instance = DCMI};
+#if (OMV_ENABLE_SENSOR_MDMA == 1)
+MDMA_HandleTypeDef DCMI_MDMA_Handle0 = {.Instance = MDMA_Channel0};
+MDMA_HandleTypeDef DCMI_MDMA_Handle1 = {.Instance = MDMA_Channel1};
```
Comment (iabdalkader) [inline src/omv/ports/stm32/sensor.c:45]:
```suggestion
static MDMA_HandleTypeDef DCMI_MDMA_Handle0 = {.Instance = MDMA_Channel0};
static MDMA_HandleTypeDef DCMI_MDMA_Handle1 = {.Instance = MDMA_Channel1};
```

File: src/omv/ports/stm32/sensor.c [hunk 2]
```diff
@@ -1130,6 +1140,42 @@ void HAL_DCMI_FrameEventCallback(DCMI_HandleTypeDef *hdcmi)
     framebuffer_get_tail(FB_NO_FLAGS);
 }
 
+#if (OMV_ENABLE_SENSOR_MDMA == 1)
+void mdma_memcpy(vbuffer_t *buffer, void *dst, void *src, int bpp, bool transposed)
```
Comment (iabdalkader) [inline src/omv/ports/stm32/sensor.c:1144]:
```suggestion
static void mdma_memcpy(vbuffer_t *buffer, void *dst, void *src, int bpp, bool transposed)
```

File: src/omv/ports/stm32/sensor.c [hunk 3]
```diff
@@ -1130,6 +1140,42 @@ void HAL_DCMI_FrameEventCallback(DCMI_HandleTypeDef *hdcmi)
...
+                   (uint32_t) dst,
+                   transposed ? bpp : (MAIN_FB()->u * bpp),
+                   transposed ? MAIN_FB()->u : 1);
+}
+#endif
+
+// If we are cropping the image by more than 1 word in width we can align the line start to
+// a word address to improve copy performance. Do not crop by more than 1 word as this will
+// result in less time between DMA transfers complete interrupts on 16-byte boundaries.
+uint32_t get_dcmi_hw_crop(uint32_t bytes_per_pixel)
```
Comment (iabdalkader) [inline src/omv/ports/stm32/sensor.c:1166]:
```suggestion
static uint32_t get_dcmi_hw_crop(uint32_t bytes_per_pixel)
```

File: src/omv/ports/stm32/sensor.c [hunk 4]
```diff
@@ -1203,99 +1248,191 @@ void DCMI_DMAConvCpltUser(uint32_t addr)
...
+            }
+            #endif
+            break;
+        default:
+            break;
+    }
+}
+
+#if (OMV_ENABLE_SENSOR_MDMA == 1)
+void mdma_config(MDMA_InitTypeDef *init, sensor_t *sensor, uint32_t bytes_per_pixel)
```
Comment (iabdalkader) [inline src/omv/ports/stm32/sensor.c:1359]:
```suggestion
static void mdma_config(MDMA_InitTypeDef *init, sensor_t *sensor, uint32_t bytes_per_pixel)
```

**Review by iabdalkader — CHANGES_REQUESTED** [review]

**Review by iabdalkader — APPROVED** [review]

**Pull request merged**